Improve caching in dindctl build and update README

pcarranzav opened the pull request:
The change in dindctl build uses automation/build.sh to pull intermediate images to improve caching. Sometimes it works to avoid
a full base image build...

Change-Type: patch
Signed-off-by: Pablo Carranza Velez <pablo@resin.io>

imrehg commented on README.md:
````diff
@@ -131,6 +130,48 @@ docker exec -it resin_supervisor_1 journalctl -f
 ```
````
missed changing this one?

pcarranzav replied:
Good catch, fixed

imrehg commented on README.md:
```diff
@@ -8,23 +8,16 @@ The Supervisor is for now a node.js program, with a subset of its functionality
 ### Set up config.json
 
-Add a `tools/dind/config.json` file from a staging device image. It should be configured for an x86 or amd64 device type so that you can push apps to it and they run properly on your computer.
+Add a `tools/dind/config.json` file from a staging or production device image. It should be configured for an x86 or amd64 device type (e.g. an Intel Nuc) so that you can push apps to it and they run properly on your computer.
```
Is the location of `config.json` still in that folder?

pcarranzav replied:
Yes, it's the same

imrehg commented on README.md:
```diff
@@ -88,13 +79,17 @@ It should look something like this:
 
```
Is the location of the `apps.json` still in that subdirectory, or also moved as `dindctl` is moved?

pcarranzav replied:
Yes, it's the same

zwalchuk commented on README.md:
```diff
@@ -8,23 +8,16 @@ The Supervisor is for now a node.js program, with a subset of its functionality
-This will build the Supervisor docker image locally. If you then run `docker images` you should see the repo/tag you
-set there.
+This process will allow you to run a development instance of the supervisor on your local computer. It is not recommended for production scenarios, but allows someone developing on the supervisor to test changes quickly.
+The supervisor is run inside a docker-in-docker container that roughly emulates its environment when running on a ResinOS device.
```
We need to make sure to capitalize Docker. Also, resinOS should not be capitalized unless it is at the beginning of a sentence.

zwalchuk commented on README.md:
```diff
@@ -8,23 +8,16 @@ The Supervisor is for now a node.js program, with a subset of its functionality
+Add a `tools/dind/config.json` file from a staging or production device image. It should be configured for an x86 or amd64 device type (e.g. an Intel Nuc) so that you can push apps to it and they run properly on your computer.
 
-Note: don't use resinstaging for production devices. This is for development purposes only. A production (resin.io) config.json should work just as fine for this local supervisor, but we also don't recommend using this in production scenarios - ResinOS is way better suited for that.
+Note: don't use staging for production devices. This is for development purposes only. A production (resin.io) config.json should work just as fine for this local supervisor, but we also don't recommend using this in production scenarios - ResinOS is way better suited for that.
```
Capitalize Don't and make resinOS lowercase. Also change "just as fine" to "just as well", and remove "way" before "better".

zwalchuk commented on README.md:
```diff
@@ -34,51 +27,49 @@ A config.json file can be obtained in several ways, for instance:
-This will mount the ./dist folder into the supervisor container, so that any changes you make can be added to the running supervisor with:
+This will mount the ./dist folder into the supervisor container and build the code before starting the instance, so that any changes you make can be added to the running supervisor with:
+
+Note: using `--mount-dist` requires a nodejs 6.x installed on your computer.
```
This Note should go after the code block

zwalchuk commented on README.md:
````diff
@@ -103,26 +98,30 @@ Make sure the config.json file doesn't have uuid, registered_at or deviceId popu
+./dindctl run --image resin/amd64-supervisor:master --preload
 ```
-This will make the docker-in-docker instance pull the image specified in apps.json before running the supervisor.
+This will make the docker-in-docker instance pull the image specified in apps.json before running the supervisor, simulating a preloaded ResinOS image (where `resin preload` has been used on the image, see [the resin CLI docs](https://docs.resin.io/tools/cli/#preload-60-image-62-)).
````
`apps.json` should be an inline code block, this is the convention we use for file names and commands.

zwalchuk commented on README.md:
```diff
@@ -131,6 +130,48 @@ docker exec -it resin_supervisor_1 journalctl -f
+## Developing with a ResinOS device
+
+If you want to test local changes (only changes to the nodejs code are supported) on a real ResinOS device, provision
+a [development OS image](https://docs.resin.io/understanding/understanding-devices/2.0.0/#dev-vs-prod-images) and power up the device. On the resin dashboard, take note of the device's IP address. Then run:
```
change "resin dashboard" to "resin.io dashboard"

pcarranzav commented:
Thanks @imrehg and @zwalchuk!
I've addressed your comments and also fixed other instances of similar issues (i.e. "resin.io VPN" instead of "resin VPN") - do those make sense?

petrosagg commented:
@pcarranzav how does git clean affect caching?

pcarranzav commented:
@petrosagg if there's additional/untracked files in the base-image folder, the caching of the corresponding layers in the docker build may be a miss, causing a rebuild of the yocto image (see here https://github.com/resin-io/resin-supervisor/blob/master/Dockerfile#L38 ) - so a git clean minimizes the chances of a rebuild there, which can be a really long build step ( > 3 hours usually).

(It's rare for us to do supervisor builds locally, but this can help make them tolerable, especially for external contributors)
